src/main.ts

```diff
@@ -0,0 +1,154 @@
+import * as cp from 'child_process'
+import * as os from 'os'
+import { AutoLanguageClient } from 'atom-languageclient'
+import * as path from 'path'
+import { CompositeDisposable } from 'atom';
+import * as pkg from '../package.json'
+import * as fs from 'fs-extra'
+import merge from 'deepmerge'
+import * as languageServer from 'ocaml-language-server'
+
+
+type DeepPartial<T> = {
+  [K in keyof T]?: DeepPartial<T[K]>
+}
+
+// hack for fix merge's types
+import config from './config'
+
+const confFile = ".atom/ide-reason.json"
+const defaultConfig = languageServer.ISettings.defaults.reason
+type Config = languageServer.ISettings['reason']
+
+const scopes = [
+  'source.reason',
+  'source.ocaml',
+  'source.re',
+  'source.ml',
+  'source.rei',
+  'source.mli',
+]
+
+function flatten1<T>(array: T[]) {
+  return ([] as T[]).concat(...array)
+}
+
+function joinEnvPath(...paths: (string | undefined)[]) {
+  return flatten1(paths)
+    .filter(Boolean)
+    .join(path.delimiter)
+}
+
+function readFileConf(dir: string, file: string, property?: string) {
+  try {
+    file = path.join(dir, file)
+    if (fs.existsSync(file)) {
+      const conf = JSON.parse(fs.readFileSync(file, 'utf8'))
+      return property ? conf[property] : conf
+    }
+  } catch (e) {
+    console.error(e)
+  }
+  return null
+}
+
+class ReasonMLLanguageClient extends AutoLanguageClient {
+  subscriptions: CompositeDisposable | null = null
+  confPerProject: DeepPartial<Config> = {}
+
+  getGrammarScopes () { return scopes }
+  getLanguageName () { return 'Reason' }
+  getServerName () { return 'ocamlmerlin' }
+  getConnectionType() { return 'ipc' as 'ipc' }
+  getRootConfigurationKey() { return 'ide-reason' }
+
+  mapConfigurationObject(config: Config) {
+    config = merge(merge({}, config), this.confPerProject)
```

Comment:
this line fails compile? (I remember once it could even fixed the `DeepPartial<T[K]>` issue.)

Reply:
It should be fixed now, I'm using a custom d.ts for `deepmerge` now.